torch/utils/cpp_extension.py

```diff
@@ -1312,6 +1318,153 @@ def load(name,
+            return True
+        return False
+
+    compiler = _get_cxx_compiler()
+
+    b_is_gcc = check_compiler_is_gcc(compiler)
+    if b_is_gcc is False:
+        return
+
+    head_file = os.path.join(_TORCH_PATH, 'include', 'torch', 'extension.h')
+    head_file_pch = os.path.join(_TORCH_PATH, 'include', 'torch', 'extension.h.gch')
+    head_file_signature = os.path.join(_TORCH_PATH, 'include', 'torch', 'extension.h.sign')
+
+    def listToString(s):
+        # initialize an empty string
+        string = ""
+        # traverse in the string
+        for element in s:
+            string += (element + ' ')
+        # return string
+        return string
+
+    def format_precompiler_header_cmd(compiler, head_file, head_file_pch, common_cflags, extra_cflags, extra_include_paths):
+        return re.sub(
+            r"[ \n]+",
+            " ",
+            f"""
+                {compiler} -x c++-header {head_file} -o {head_file_pch} {extra_include_paths} {extra_cflags} {common_cflags}
+            """,
+        ).strip()
+
+    def command_to_signature(cmd):
+        signature = cmd.replace(' ', '_')
+        return signature
+
+    def check_pch_signature_in_file(file_path, signature):
+        b_exist = os.path.isfile(file_path)
+        if b_exist is False:
+            return False
+
+        with open(file_path) as file:
+            # read all content of a file
+            content = file.read()
+            # check if string present in a file
+            if signature == content:
+                return True
+            else:
+                return False
+
+    def _create_if_not_exist(path_dir):
+        if not os.path.exists(path_dir):
+            try:
+                Path(path_dir).mkdir(parents=True, exist_ok=True)
+            except OSError as exc:  # Guard against race condition
+                if exc.errno != errno.EEXIST:
+                    raise RuntimeError("Fail to create path {}".format(path_dir))
+
+    def write_pch_signature_to_file(file_path, pch_sign):
+        _create_if_not_exist(os.path.dirname(file_path))
+        with open(file_path, "w") as f:
+            f.write(pch_sign)
+            f.close()
+
+    def build_precompile_header(pch_cmd):
+        try:
+            subprocess.check_output(pch_cmd, shell=True, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            from .._inductor.exc import CppCompileError
+            raise CppCompileError(pch_cmd, e.output) from e
+
+    extra_cflags_str = listToString(extra_cflags)
+    extra_include_paths_str = listToString(extra_include_paths)
+
+    common_cflags = []
+    if not is_standalone:
+        common_cflags += ['-DTORCH_API_INCLUDE_EXTENSION_H']
+
+    common_cflags += ['-std=c++17', '-fPIC']
+    common_cflags += [f"{x}" for x in _get_pybind11_abi_build_flags()]
+    common_cflags += [f"{x}" for x in _get_glibcxx_abi_build_flags()]
+    common_cflags_str = listToString(common_cflags)
```

Comment:
How is cuda/hip build supported? They need extra cflags.

Reply:
@jgong5 
1. We only make PCH for extension.h. CUDA and HIP will also get benifit.
2. If use CUDA/HIP, cflag will changed. and then signature checking will re-make PCH file.

Reply:
But when you generate the pch, CUDA/HIP headers are not incorporated, right? How can you generate the right PCH for them?

Reply:
1. cuda.h is a small content header. It is not much benifit to make PCH.
2. cuda_runtime.h may not locate in pytorch.
3. according to my profilng, the most time cosumerd in extension.h and its Python.h
4. If we find cuda also need PCH, we can figure out a solution later.
<img width="496" alt="image" src="https://github.com/pytorch/pytorch/assets/8433590/631105c0-b7d1-40a8-8e06-9f78ba780a43">